src/main/java/com/microsoft/aad/msal4j/HttpHelper.java

```diff
@@ -42,17 +65,66 @@ static IHttpResponse executeHttpRequest(final HttpRequest httpRequest,
                 HttpHelper.verifyReturnedCorrelationId(httpRequest, httpResponse);
             }
         }
+        processThrottlingInstructions(httpResponse, requestContext);
+
         return httpResponse;
     }
 
-    private static void addRequestInfoToTelemetry(final HttpRequest httpRequest, HttpEvent httpEvent){
-        try{
+    private static void checkForThrottling(RequestContext requestContext) {
+        if (requestContext.clientApplication() instanceof PublicClientApplication &&
+                requestContext.apiParameters() != null) {
+            String requestThumbprint = getRequestThumbprint(requestContext);
+
+            long retryInMs = ThrottlingCache.retryInMs(requestThumbprint);
+
+            if (retryInMs > 0) {
+                throw new MsalThrottlingException(retryInMs);
+            }
+        }
+    }
+
+    private static void processThrottlingInstructions(IHttpResponse httpResponse, RequestContext requestContext) {
+        if (requestContext.clientApplication() instanceof PublicClientApplication) {
+            Long expirationTimestamp = null;
+
+            Integer retryAfterHeaderVal = getRetryAfterHeader(httpResponse);
+            if (retryAfterHeaderVal != null) {
+                expirationTimestamp = System.currentTimeMillis() + retryAfterHeaderVal * 1000;
+            } else if (httpResponse.statusCode() == 429 ||
+                    (httpResponse.statusCode() >= 500 && httpResponse.statusCode() < 600)) {
```

Comment:
We want to throttle every time we get any 500 status code back?

Reply:
according to design doc yes, but only requests with same thumbprint.
I see your concern, generally speaking 5xx errors are retriable. 
It would be better to add some retry logic around this, and throttle only if all retries failed